### ISSUE TITLE
KeyError in set_duplicate_name_id()

### ISSUE THREAD

**Issue by kovidgoyal** [sphinx-doc/sphinx#5150]
Traceback:
```
# Sphinx version: 1.7.5
# Python version: 2.7.15 (CPython)
# Docutils version: 0.14 
# Jinja2 version: 2.10
# Last messages:
#   creating lrfviewer
#   creating lrs2lrf
#   creating web2disk
#   creating template_ref
#   building [mo]: targets for 0 po files that are out of date
#   building [html]: targets for 29 source files that are out of date
#   updating environment:
#   47 added, 0 changed, 0 removed
#   reading sources... [  2%] catalogs
#   reading sources... [  4%] conversion
# Loaded extensions:
#   sidebar_toc (unknown version) from /home/kovid/work/calibre/manual/sidebar_toc.pyc
#   sphinx.ext.viewcode (1.7.5) from /usr/lib/python2.7/site-packages/sphinx/ext/viewcode.pyc
#   alabaster (0.7.11) from /usr/lib/python2.7/site-packages/alabaster/__init__.py
#   custom (unknown version) from /home/kovid/work/calibre/manual/custom.pyc
#   sphinx.ext.autodoc (1.7.5) from /usr/lib/python2.7/site-packages/sphinx/ext/autodoc/__init__.pyc
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/sphinx/cmdline.py", line 304, in main
    app.build(args.force_all, filenames)
  File "/usr/lib/python2.7/site-packages/sphinx/application.py", line 331, in build
    self.builder.build_update()
  File "/usr/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 342, in build_update
    'out of date' % len(to_build))
  File "/usr/lib/python2.7/site-packages/sphinx/builders/__init__.py", line 355, in build
    updated_docnames = set(self.env.update(self.config, self.srcdir, self.doctreedir))
  File "/usr/lib/python2.7/site-packages/sphinx/environment/__init__.py", line 565, in update
    self._read_serial(docnames, self.app)
  File "/usr/lib/python2.7/site-packages/sphinx/environment/__init__.py", line 584, in _read_serial
    self.read_doc(docname, app)
  File "/usr/lib/python2.7/site-packages/sphinx/environment/__init__.py", line 659, in read_doc
    doctree = read_doc(self.app, self, self.doc2path(docname))
  File "/usr/lib/python2.7/site-packages/sphinx/io.py", line 294, in read_doc
    pub.publish()
  File "/usr/lib/python2.7/site-packages/docutils/core.py", line 218, in publish
    self.apply_transforms()
  File "/usr/lib/python2.7/site-packages/docutils/core.py", line 199, in apply_transforms
    self.document.transformer.apply_transforms()
  File "/usr/lib/python2.7/site-packages/sphinx/transforms/__init__.py", line 96, in apply_transforms
    Transformer.apply_transforms(self)
  File "/usr/lib/python2.7/site-packages/docutils/transforms/__init__.py", line 171, in apply_transforms
    transform.apply(**kwargs)
  File "/usr/lib/python2.7/site-packages/sphinx/transforms/i18n.py", line 182, in apply
    self.document.note_implicit_target(section_node)
  File "/usr/lib/python2.7/site-packages/docutils/nodes.py", line 1383, in note_implicit_target
    self.set_name_id_map(target, id, msgnode, explicit=None)
  File "/usr/lib/python2.7/site-packages/docutils/nodes.py", line 1330, in set_name_id_map
    self.set_duplicate_name_id(node, id, name, msgnode, explicit)
  File "/usr/lib/python2.7/site-packages/docutils/nodes.py", line 1367, in set_duplicate_name_id
    old_node = self.ids[old_id]
KeyError: 'contents'
```

The command line used:
```
sphinx-build2 -D language=cs -b html -q -w /t/cs/html/sphinx-build-warnings.txt -t online -d /t/cs/doctrees /home/kovid/work/calibre/manual /t/cs/html
```

This happens only when running sphinx with i18n. Sadly it is in a quite large/complex project and I have not been able to create a minimal test case. sphinx version is 1.7.5

**Comment by tk0miya** [comment]
Thank you for reporting. Absolutely this is a bug of i18n module. It would be nice if you'll make smaller project. I'll also try to reproduce it on my local later.

**Comment by kovidgoyal** [comment]
Yeah, I tried creating a smaller project to reproduce I failed :( For the moment I worked around it by patching docutils to ignore that error.

**Comment by tk0miya** [comment]
Any updates?

**Comment by kovidgoyal** [comment]
What update are you hoping for? As I said, I failed to create a small project to duplicate the error

**Comment by tk0miya** [comment]
Oh, sorry. I'd misread. I've also failed to reproduce...

**Comment by kovidgoyal** [comment]
Note that simply ignoring the KeyError exception in transforms/i18n.py allows the build to complete. I have no idea what other side-effects it might have.

**Comment by tk0miya** [comment]
I can't answer ignoring is best way for this case. So I'll continue to reproduce the error.

**Comment by jpmckinney** [comment]
Is this fixed in #6718?

**Comment by kovidgoyal** [comment]
Not sure, when the arch package for sphinx is updated to 2.3.1 I will
test and let you know.

**Comment by eli-schwartz** [comment]
I commented out the monkey_patch_docutils() function in calibre's manual/custom.py file (currently being used to except and pass this KeyError).

With this change, calibre's manual does not build correctly (as per the initial report) using the command line:
```
sphinx-build2 -D language=el -b man -q -Q -d /home/eschwartz/git/calibre/man-pages/doctrees /home/eschwartz/git/calibre/manual /home/eschwartz/git/calibre/man-pages/el
```

Then I hotpatched /usr/lib/python2.7/site-packages/sphinx/transforms/i18n.py with the fix in #6731 and retried sphinx-build2, and the manual succeeded in building. So I'd say this is indeed solved.
